File: C++/1654_랜선자르기.cpp
#include <iostream>
using namespace std;

bool isPossible(int arr[], long long int mid, int K, int N);

int main() {
	int K, N;
	cin >> K >> N;
	
	int arr[K]; // 랜선들의 길이를 담을 배열
	long long int right = 0, left = 1; // 최대 길이의 랜선을 기준으로 수행할 것 right는 가장 긴 랜선의 길이가 될 것임
	for(long long int i = 0; i < K; i++){
		cin >> arr[i];
		
		if(right < arr[i]){
			right = arr[i];
		}
	}
	
	long long int result = 0; // 최대 길이를 담을 변수
	while(right >= left){ // 이분탐색 진행
		long long int mid = (right+left)/2;
		bool flag = isPossible(arr,mid,K,N); // N개의 랜선이 만들어졌는지
		if(flag == true){ // N개의 랜선이 만들어졌다면
			left = mid+1; // 랜선의 크기를 키워서 탐색(큰쪽 탐색)
			result = mid; // mid 값이 랜선의 길이가 되는데, 만약 N개의 랜선이 만들어졌다면 result에 일단 대입(계속 대입해도 되는 이유: N개의 랜선이 만들어졌다면 이전의 mid값보다 작아질 일 없음)
		}
		else{ // N개의 랜선이 만들어지지 않았다면
			right = mid-1; // 랜선의 크기를 줄여서 탐색(작은쪽 탐색)
		}
	}
	
	cout << result;

	return 0;
}

bool isPossible(int arr[], long long int mid, int K, int N){ // N개의 랜선이 만들어졌는지 확인하는 함수
	int tmp = 0;
	for(int i = 0; i < K; i++){
		tmp += arr[i]/mid; // 랜선이 몇 개 만들어지는지 count
	}
	
	if(tmp >= N){ // N개보다 많이 만들어져도 되기 때문에 현재 만들어진 랜선이 목표하는 랜선의 수보다 크거나 같다면 true를 return
		return true;
	}
	
	return false; // 아니라면 false를 return
}

File: Python/1654_랜선자르기.py
def isPossible(K, N, arr, mid): # N개 이상의 랜선이 만들어지는지 확인하기 위한 함수
	tmp = 0
	for a in arr:
		tmp += int(a/mid) # mid: 현재 랜선의 길이 / 현재 랜선 길이로 설정되었다면 몇 개의 랜선이 만들어지는지 count
	
	if tmp >= N: # N개 이상의 랜선이 만들어졌다면, true를 return
		return True
	
	return False # N개보다 적은 랜선이 만들어졌다면, false를 return
	

if __name__ == '__main__':
	K, N = list(map(int,input().split()))
	
	arr = [int(input()) for _ in range(K)]
	
	left = 1
	right = max(arr)
	
	result = 0
	while left <= right: # 이분 탐색 시작
		mid = int((left+right)/2) # mid 값이 랜선의 최대 길이를 의미
		if isPossible(K,N,arr,mid): # N개 이상의 랜선이 만들어졌다면
			left = mid + 1 # 현재 랜선의 길이보다 긴 랜선의 범위 탐색(이분 탐색 재진행)
			result = mid # N개 이상의 랜선이 만들어졌다면, 이전 mid 값보다 작은 mid값일 수가 없음, 따라서 result 계속 갱신해주기
		else: # N개의 랜선이 만들어지지 않았다면
			right = mid - 1 # 현재 랜선의 길이보다 작은 랜선의 범위 탐색
			
	print(result)
